laserchicken/feature_extractor/test_density_extractor.py

```diff
@@ -0,0 +1,188 @@
+import os
+import unittest
+
+import random
+import numpy as np
+
+from laserchicken import keys, read_las, utils
+from laserchicken.compute_neighbors import compute_neighborhoods
+from laserchicken.volume_specification import Sphere, InfiniteCylinder
+from laserchicken.feature_extractor.density_feature_extractor import PointDensityFeatureExtractor
+
+
+class TestDensityFeatureExtractorSphere(unittest.TestCase):
+    """Test density extractor on artificial spherical data."""
+
+    point_cloud = None
+
+    def test_sphere_index(self):
+        """Compute the density for a sphere given as index of the source pc."""
+        neighbors_index = compute_neighborhoods(self.point_cloud,
+                                                self.targetpc,
+                                                self.sphere)
+
+        extractor = PointDensityFeatureExtractor()
+        for index in neighbors_index:
+            d = extractor.extract(self.point_cloud, index, None, None, self.sphere)
+            self.assertEqual(d, self.voldens)
+
+    def _get_central_point(self):
+        """Get the central point."""
+        return utils.copy_pointcloud(self.point_cloud, [0])
+
+    def _get_points_on_sphere(self):
+        """Generate a pc of points regularly positionned on a two spheres of radius 1 and 2."""
+        nteta, nphi = 11, 11
+        self.teta = np.linspace(0.1, 2 * np.pi, nteta)
+        self.phi = np.linspace(0.1, np.pi, nphi)
+        self.radius = [1., 2.]
+        self.points_per_sphere = nteta * nphi
+        self.xyz = []
+        self.xyz.append([0., 0., 0.])
+        for r in self.radius:
+            for t in self.teta:
+                for p in self.phi:
+                    x = r * np.cos(t) * np.sin(p)
+                    y = r * np.sin(t) * np.sin(p)
+                    z = r * np.cos(p)
+                    self.xyz.append([x, y, z])
+
+        self.xyz = np.array(self.xyz)
+        self.point_cloud = {keys.point: {'x': {'type': 'double', 'data': self.xyz[:, 0]},
+                           'y': {'type': 'double', 'data': self.xyz[:, 1]},
+                           'z': {'type': 'double', 'data': self.xyz[:, 2]}}}
+
+    def setUp(self):
+        """Set up the test."""
+        # get the points
+        self._get_points_on_sphere()
+
+        # get the central point as targetpc
+        self.targetpc = self._get_central_point()
+
+        # get the sphere
+        self.sphere = Sphere(np.mean(self.radius))
+
+        # get the theoretical value +1 for central point
+        npts = self.points_per_sphere + 1
+        self.voldens = npts / self.sphere.calculate_volume()
+
+    def tearDown(self):
+        pass
+
+
+class TestDensityFeatureExtractorCylinder(unittest.TestCase):
+    """Test density extractor on artificial cylindric data."""
+
+    point_cloud = None
+
```

Comment:
Looks rather complicated to just test the density calculation. Actually I don't really get how that this is actually something that should pass.